### PR TITLE
Fix #859: ARG not support in FROM

### LINKED ISSUE #859: ARG not support in FROM
### Description

Trying to build a `Dockerfile` like the following fails when using the maven plugin, but is supported and succeeds when building from the command line.

https://docs.docker.com/engine/reference/builder/#from

```
ARG VERSION:latest
FROM something:$VERSION
```

```
[ERROR] Failed to execute goal io.fabric8:docker-maven-plugin:0.22.1:build (build-docker-image) on project docker: Execution build-docker-image of goal io.fabric8:docker-maven-plugin:0.22.1:build failed: Given Docker name 'something:$VERSION' is invalid:
[ERROR] * tag part '$VERSION' doesn't match allowed pattern '^[\w][\w.-]{0,127}$'
[ERROR] See http://bit.ly/docker_image_fmt for more details
[ERROR] -> [Help 1]
```

VS when run from the command line

```
Sending build context to Docker daemon  482.7MB
Step 1/14 : ARG VERSION=latest
 ---> 
Step 2/14 : FROM something:$VERSION
 ---> bd7be3f0d8c6
...........
...........
```

### Info
* Maven version (`mvn -v`) :
```
Apache Maven 3.3.9 (bb52d8502b132ec0a5a3f4c09453c07478323dc5; 2015-11-10T08:41:47-08:00)
Maven home: /opt/apache-maven-3.3.9
Java version: 1.8.0_112, vendor: Oracle Corporation
Java home: /Library/Java/JavaVirtualMachines/jdk1.8.0_112.jdk/Contents/Home/jre
Default locale: en_US, platform encoding: UTF-8
OS name: "mac os x", version: "10.12.6", arch: "x86_64", family: "mac"
```
* Docker version (`docker version`):
```
Client:
 Version:      17.06.2-ce
 API version:  1.30
 Go version:   go1.8.3
 Git commit:   cec0b72
 Built:        Tue Sep  5 20:12:06 2017
 OS/Arch:      darwin/amd64

Server:
 Version:      17.06.2-ce
 API version:  1.30 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   cec0b72
 Built:        Tue Sep  5 19:59:19 2017
 OS/Arch:      linux/amd64
 Experimental: true
```

### PR DESCRIPTION
**Pull request by rohanKanojia**
Fix #859

### CODE REVIEW

File: doc/changelog.md [hunk 1]
```diff
@@ -8,6 +8,7 @@
   - Add docker:build support for 'network' option #1030
   - Failure referencing a previous staged image in FROM clause #1264
   - Treat bridged and default network mode the same (#1234)
+  - ARG not support in FROM (#859)
```
Comment (rhuss) [inline doc/changelog.md]:
```suggestion
  - Support `ARG ` in `FROM` (#859)
```

File: src/main/java/io/fabric8/maven/docker/util/DockerFileUtil.java [hunk 2]
```diff
@@ -55,23 +56,47 @@ private DockerFileUtil() {}
...
+     * @return HashMap of arguments or empty collection if none is found
+     * @throws IOException in case of any problem reading file
+     */
+    public static Map<String, String> extractArgs(File dockerFile, FixedStringSearchInterpolator interpolator) throws IOException {
+        List<String[]> argLines = extractLines(dockerFile, "ARG", interpolator);
+        Map<String, String> result = new HashMap<>();
+        for (String[] argLine : argLines) {
+            if (argLine.length > 1) {
+                if (argLine[1].contains(":")) {
+                    result.put(argLine[1].split(":")[0], argLine[1].split(":")[1]);
```
Comment (rhuss) [inline src/main/java/io/fabric8/maven/docker/util/DockerFileUtil.java]: * You should only split once and reuse the result
* You should split into exactly two parts, with the second parts potentially having `:`, too. I.e. use the split function which takes the max. nr of resulting split elements as argument.

File: src/main/java/io/fabric8/maven/docker/util/DockerFileUtil.java [hunk 3]
```diff
@@ -55,23 +56,47 @@ private DockerFileUtil() {}
...
+                } else if (argLine[1].contains("=")) {
+                    result.put(argLine[1].split("=")[0], argLine[1].split("=")[1]);
```
Comment (rhuss) [inline src/main/java/io/fabric8/maven/docker/util/DockerFileUtil.java]: You can put the split on ':' and '=' into a function for reuse, e.g. `addArgWitValue(result, "=", argLine[1])`

**Comment by rohanKanojia** [comment]
I was trying to do the changes but I noticed one thing, I'm testing using this Dockerfile(to test `FROM $IMAGENAME` scenario):
```
FROM openjdk:jre

ARG version=0.31-SNAPSHOT
ARG jar_file=target/demp-sample-zero-config-$version.jar
ARG FULLIMAGENAME=busybox

ADD $jar_file /tmp/zero-config.jar
CMD java -cp /tmp/zero-config.jar HelloWorld
FROM $FULLIMAGENAME
```
But docker doesn't seem to be able to process this:
```
~/work/repos/docker-maven-plugin/samples/zero-config : $ docker build .
Sending build context to Docker daemon  45.06kB
Step 1/7 : FROM openjdk:jre
 ---> 6b23d41384f9
Step 2/7 : ARG version=0.31-SNAPSHOT
 ---> Using cache
 ---> de78be110994
Step 3/7 : ARG jar_file=target/demp-sample-zero-config-$version.jar
 ---> Using cache
 ---> a8268726a072
Step 4/7 : ARG FULLIMAGENAME=busybox
 ---> Using cache
 ---> f802a9b7666c
Step 5/7 : ADD $jar_file /tmp/zero-config.jar
 ---> Using cache
 ---> 15a65328a8aa
Step 6/7 : CMD java -cp /tmp/zero-config.jar HelloWorld
 ---> Using cache
 ---> dd5d7d159fea
Step 7/7 : FROM $FULLIMAGENAME
base name ($FULLIMAGENAME) should not be blank
```

Were you talking about this case? I'm not sure whether I've understood your comment correctly or not. I'm on 1.40 API version at the moment. 
```
~/work/repos/docker-maven-plugin/samples/zero-config : $ docker version
Client: Docker Engine - Community
 Version:           19.03.4
 API version:       1.40
 Go version:        go1.12.10
 Git commit:        9013bf583a
 Built:             Fri Oct 18 15:52:46 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.4
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.12.10
  Git commit:       9013bf583a
  Built:            Fri Oct 18 15:51:19 2019
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.2.10
  GitCommit:        b34a5c8af56e510852c35414db4c1f4fa6172339
 runc:
  Version:          1.0.0-rc8+dev
  GitCommit:        3e425f80a8c931f88e6d94a8c831b9d5aa481657
 docker-init:
  Version:          0.18.0
```

**Comment by rohanKanojia** [comment]
Sorry, it seems to be working in this case:
```
ARG FULLIMAGENAME=busybox:latest
FROM $FULLIMAGENAME
```

File: src/main/java/io/fabric8/maven/docker/util/DockerFileUtil.java [hunk 4]
```diff
@@ -137,6 +162,45 @@ public static FixedStringSearchInterpolator createInterpolator(MojoParameters pa
...
+    /**
+     * Iterates through the argument string and splits it on first delimiter found(whether it's `:` or `=`.
+     *
+     * @param argString argument string
+     * @return a pair containing argument key and value
+     */
+    private static AbstractMap.SimpleEntry<String, String> resolveArgsFromArgString(String argString) {
+        for (int index = 0; index < argString.length(); index++) {
+            if (argString.charAt(index) == ':' || argString.charAt(index) == '=') {
+                return new AbstractMap.SimpleEntry<>(argString.substring(0, index), argString.substring(index+1));
```
[the remark this answers is not in the document]
Comment (rhuss) [reply src/main/java/io/fabric8/maven/docker/util/DockerFileUtil.java]: What if index is the lasst char in the string ? Wouldn't it be better to use String.split() with a regexp like `:|=` ?

File: src/main/java/io/fabric8/maven/docker/util/DockerFileUtil.java [hunk 5]
```diff
@@ -55,23 +57,46 @@ private DockerFileUtil() {}
...
+     * @param interpolator interpolator for replacing properties
+     * @return HashMap of arguments or empty collection if none is found
+     * @throws IOException in case of any problem reading file
+     */
+    public static Map<String, String> extractArgs(File dockerFile, FixedStringSearchInterpolator interpolator) throws IOException {
+        List<String[]> argLines = extractLines(dockerFile, "ARG", interpolator);
+        Map<String, String> result = new HashMap<>();
+        for (String[] argLine : argLines) {
+            if (argLine.length > 1) {
+                AbstractMap.SimpleEntry<String, String> resolvedArg = resolveArgsFromArgString(argLine[1]);
```
Comment (rhuss) [inline src/main/java/io/fabric8/maven/docker/util/DockerFileUtil.java]: I would call the method `updateMapWithArgValue()` and feed in the map as argument. That allows you to avoid to generate temporary, short-lived obkects like `resolvedArg`

File: src/test/java/io/fabric8/maven/docker/util/DockerFileUtilTest.java [hunk 6]
```diff
@@ -69,6 +69,17 @@ public void testMultiStageNamed() throws Exception {
...
+    public void testMultiStageWithArgs() throws Exception {
+        File toTest = copyToTempDir("Dockerfile_multi_stage_with_args");
+        Iterator<String> fromClauses = DockerFileUtil.extractBaseImages(
+                toTest, FixedStringSearchInterpolator.create()).iterator();
+
+        assertEquals("fabric8/s2i-java:latest", fromClauses.next());
+        assertEquals("busybox:latest", fromClauses.next());
+        assertEquals(false, fromClauses.hasNext());
+    }
+
```
Comment (rhuss) [inline src/test/java/io/fabric8/maven/docker/util/DockerFileUtilTest.java]: I would love to see more tests directly for the parsing methods with many corner cases. No need to create test Dockerfiles, but calling `extractArgs()` and `resolveImageTagFromArgs()` in a loop with different inputs (you could make the methods package private). But we need much more tests about different cases (like with ARG values, containing spaces, if this is allowed. Also check the Docker spec for the possible ARG variations).

**Comment by rohanKanojia** [comment]
@rhuss : Hi, Sorry I forgot to update this. I will update this around coming weekend

File: src/main/java/io/fabric8/maven/docker/util/DockerFileUtil.java [hunk 7]
```diff
@@ -55,21 +55,38 @@ private DockerFileUtil() {}
...
+        return new ArrayList<>(result);
+    }
+
+    /**
+     * Extract Args from dockerfile. All lines containing ARG is taken.
+     *
+     * @param argLines list of lines containing arguments
+     * @return HashMap of arguments or empty collection if none is found
+     */
+    public static Map<String, String> extractArgs(List<String[]> argLines) {
```
Comment (rhuss) [inline src/main/java/io/fabric8/maven/docker/util/DockerFileUtil.java:82]: Maybe you could inline the usage of `extractLines` ? (e..g change the signature to `extractArgs(String dockerfile, FIxedStringSearchInterpolator interpolator)` and call extract Lines internally)

File: src/test/java/io/fabric8/maven/docker/util/DockerFileUtilTest.java [hunk 8]
```diff
@@ -66,7 +64,57 @@ public void testMultiStageNamed() throws Exception {
...
+        Iterator<String> fromClauses = DockerFileUtil.extractBaseImages(
+                toTest, FixedStringSearchInterpolator.create()).iterator();
+
+        assertEquals("fabric8/s2i-java:latest", fromClauses.next());
+        assertEquals("busybox:latest", fromClauses.next());
+        assertFalse(fromClauses.hasNext());
+    }
+
+    @Test
+    public void testExtractArgsFromDockerfile() {
```
Comment (rhuss) [inline src/test/java/io/fabric8/maven/docker/util/DockerFileUtilTest.java:82]: 👍

**Comment by dsmiley** [comment]
This issue that this PR fixes is important; I hope the PR can be merged and released.

**Comment by rhuss** [comment]
Thanks, let's merge that now.
